### ISSUE TITLE
Minimal macOS menu bar extra's app with SwiftUI

### ISSUE THREAD

**Issue by khorbushko** [khorbushko/khorbushko.github.io#40]

**Comment by khorbushk** [comment]
<img width="896" alt="Знімок екрана 2022-06-22 о 20 14 28" src="https://user-images.githubusercontent.com/96495602/175097888-9a64684e-9f25-4565-8c17-b4033bb1f1ea.png">

**Comment by marc-medley** [comment]
A comment regarding the addition of `LSUIElement` in ["Minimal macOS menu bar extra's app with SwiftUI"](https://khorbushko.github.io/article/2021/04/30/minimal-macOS-menu-bar-extra's-app-with-SwiftUI.html)…

``` xml
<key>LSUIElement</key>
<true/>
```

If `Info.plist` is manually added and edited, then the project build settings needs to be updated to _not use_ the `GENERATE_INFOPLIST_FILE ` "Generated Info.plist File" AND to _know where to find_ the manually added `INFOPLIST_FILE` "Info.plist File".

![Packaging_plist](https://user-images.githubusercontent.com/15098151/170603517-d76709c5-9481-4372-ae55-06dbaabdc6d4.png)

Source page: [2021-04-30-minimal-macOS-menu-bar-extra's-app-with-SwiftUI.md](https://github.com/khorbushko/khorbushko.github.io/blob/master/_posts/2021-04-30-minimal-macOS-menu-bar-extra's-app-with-SwiftUI.md)

**Comment by khorbushk** [comment]
post updated, thank @mark-medley

**Comment by liemeldert** [comment]
Hey there!
I know you've posted this a while ago but I just wanted to bring your attention to the addition of MenuBarExtras into native SwiftUI. [Here's their docs](https://developer.apple.com/documentation/swiftui/menubarextra), perhaps add a notice at the top that there's a new way to go about this?

**Comment by khorbushko** [comment]
@liemeldert good catch, sorry - due to the [war](https://war.ukraine.ua/) I have no time to properly manage this blog. Post updated

**Comment by dieb** [comment]
Great blog post, thank you.

@liemeldert I've used MenuBarExtras and I think it may not be ready for prime-time yet. Some issues:

- When using `.menuBarExtraStyle(.menu)` your ability to style the menu is severely limited. HStacks, VStacks and others are ignored. Works well for a menu of simple commands and menus.
- When using `.menuBarExtraStyle(.window)` you can use any SwiftUI you want, but it will not automatically close the menu when you click outside. I couldn't find docs or examples on how to do this, but found a few workarounds [here](https://damian.fyi/swift/2022/12/29/detecting-when-a-swiftui-menubarextra-with-window-style-is-opened.html) and [here](https://github.com/orchetect/MenuBarExtraAccess)
- If you need access to the status item, NSWindow, and programatically hide/close the menu, you need a workaround or a lib like [MenuBarExtraAccess](https://github.com/orchetect/MenuBarExtraAccess)
- If you make use of some implicit defaults like system padding `.padding()` in your views, you may need to override them manually to match the ones actually used in the menu bar. I also remember the translucency in the window did not match macOS Sonoma by default, you have to do something custom.
- I tried using [MenuBarExtraAcess](https://github.com/orchetect/MenuBarExtraAccess) and [MacControlCenterUI](https://github.com/orchetect/MacControlCenterUI) that wrap MenuBarExtra and they work reasonably well, but found some issues perhaps with the way I'm using it.

Bottom line I ended up switching back to NSMenu and NSMenuItems hooked to views similar to the blog post, so thanks again for it. I made the following modification to help with proper resizing:

```swift
let hostingView = NSHostingView(rootView: modifiedView)
hostingView.translatesAutoresizingMaskIntoConstraints = false
```
